tutorials/haskell/mustache-templates/tutorial.md

```diff
@@ -0,0 +1,404 @@
+them selected:
+
+```haskell
+-- | Mustache template as name of “top-level” template and a collection of
+-- all available templates (partials).
+--
+-- 'Template' is a 'Semigroup'. This means that you can combine 'Template's
+-- (and their caches) using the ('<>') operator, the resulting 'Template'
+-- will have the same currently selected template as the left one. Union of
+-- caches is also left-biased.
+
+data Template = Template
+  { templateActual :: PName
+    -- ^ Name of currently “selected” template (top-level one).
+  , templateCache  :: Map PName [Node]
+    -- ^ Collection of all templates that are available for interpolation
+    -- (as partials). The top-level one is also contained here and the
+    -- “focus” can be switched easily by modifying 'templateActual'.
+  } deriving (Eq, Ord, Show, Data, Typeable, Generic)
+```
+
+When working with a `Template`, it's only possible to use partials  that are
+in `templateCache`, thus the main difference between compiling functions is
+where you get your template(s) and what you will have in `templateCache`.
+Thus there are three different ways to get a `Template`:
+
+1. From lazy `Text` with `compileMustacheText`. The function takes a `PName`
+   and actual template source. The function returns `Either` parse error or
+   `Template` with cache consisting of only one `Template`.
+
+2. From a single `File` with `compileMustacheFile`. This one only takes path
+   to Mustache file the resulting template won't be able to use partials
+   unless you combine several templates into one (see below).
+
+3. From a directory with `compileMustacheDir`. This function reads all
+   templates (files ending with `.mustache` extension) and puts them into
+   the cache, selecting one of them as main (you specify which). The
+   resulting `Template` can use partials that were present in that
+   directory.
+
+So to use partials, we need to use `compileMustacheDir` *or* combine several
+templates into one using `(<>)` method of `Semigroup` type class. Being an
+instance of the `Semigroup` type class means that you can always to two
+`Template`s and get their combination which will also be a `Template`. This
+is an incredible property, one of numerous examples how Haskell keeps
+systems “flat”.
+
+Let's start to build practical stuff to see how things play together. As an
+example of a task that could involve Mustache templates, we will be
+generating a
+[CUE sheet](https://en.wikipedia.org/wiki/Cue_sheet_(computing)). First of
+all, we're going to need some imports:
+
+```haskell
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
+
+module Main (main) where
+
+import Data.Aeson
+import Data.Semigroup ((<>))
+import Data.Text (Text)
+import GHC.Generics
+import Text.Mustache
+import qualified Text.Mustache.Compile.TH as TH
+import qualified Data.Text.Lazy.IO as TIO
+```
+
+Then we will need to represent data somehow, for that let's define the
```

Comment:
@mrkkrp "Then we will need to represent data somehow. For that, let's define..."